onnx/defs/traditionalml/defs.cc

```diff
@@ -298,119 +298,129 @@ ONNX_ML_OPERATOR_SET_SCHEMA(
+    When 'values_as_tensor' is used, a default value must be specified.<br>
+    Float keys with value 'NaN' match any input 'NaN' value regardless of bit
+    value. If a key is repeated, the last key takes precedence.
 )DOC";
 
 ONNX_ML_OPERATOR_SET_SCHEMA(
     LabelEncoder,
-    2,
+    4,
     OpSchema()
-        .SetDoc(LabelEncoder_ver2_doc)
-        .Input(0, "X", "Input data. It can be either tensor or scalar.", "T1")
-        .Output(0, "Y", "Output data.", "T2")
+        .SetDoc(LabelEncoder_ver4_doc)
+        .Input(0, "X", "Input data. It must have the same element type as the keys_* attribute set.", "T1")
+        .Output(0, "Y", "Output data. This tensor's element type is based on the values_* attribute set.", "T2")
         .TypeConstraint(
             "T1",
-            {"tensor(string)", "tensor(int64)", "tensor(float)"},
+            {"tensor(string)", "tensor(int64)", "tensor(float)", "tensor(int32)", "tensor(int16)"},
             "The input type is a tensor of any shape.")
         .TypeConstraint(
             "T2",
-            {"tensor(string)", "tensor(int64)", "tensor(float)"},
+            {"tensor(string)", "tensor(int64)", "tensor(float)", "tensor(int32)", "tensor(int16)"},
             "Output type is determined by the specified 'values_*' attribute.")
         .Attr(
-            "keys_strings",
-            "A list of strings. One and only one of 'keys_*'s should be set.",
-            AttributeProto::STRINGS,
+            "keys_as_tensor",
+            "Keys encoded as a 1D tensor. One and only one of 'keys_*'s should be set.",
+            AttributeProto::TENSOR,
             OPTIONAL_VALUE)
+        .Attr("keys_strings", "A list of strings.", AttributeProto::STRINGS, OPTIONAL_VALUE)
         .Attr("keys_int64s", "A list of ints.", AttributeProto::INTS, OPTIONAL_VALUE)
         .Attr("keys_floats", "A list of floats.", AttributeProto::FLOATS, OPTIONAL_VALUE)
         .Attr(
-            "values_strings",
-            "A list of strings. One and only one of 'value_*'s should be set.",
-            AttributeProto::STRINGS,
+            "values_as_tensor",
+            "Values encoded as a 1D tensor. One and only one of 'values_*'s should be set.",
+            AttributeProto::TENSOR,
             OPTIONAL_VALUE)
+        .Attr("values_strings", "A list of strings.", AttributeProto::STRINGS, OPTIONAL_VALUE)
         .Attr("values_int64s", "A list of ints.", AttributeProto::INTS, OPTIONAL_VALUE)
         .Attr("values_floats", "A list of floats.", AttributeProto::FLOATS, OPTIONAL_VALUE)
         .Attr("default_string", "A string.", AttributeProto::STRING, std::string("_Unused"))
         .Attr("default_int64", "An integer.", AttributeProto::INT, static_cast<int64_t>(-1))
         .Attr("default_float", "A float.", AttributeProto::FLOAT, -0.f)
+        .Attr(
+            "default_as_tensor",
+            "A default tensor. Must be set if values_as_tensor is set, optional otherwise.",
```

Comment:
Why is it mandatory if values_as_tensor is set?

Reply:
See https://github.com/onnx/onnx/issues/5467. Since we cannot condition the default value of the tensor based on the values_* element type we cannot really describe a default value. We need a `default_as_tensor` in the first place because the element types covered by values_as_tensor is wider than that covered by default_{string|int64|float}.

Reply:
We cannot currently cleanly infer this as described in https://github.com/onnx/onnx/issues/5467. The attribute is necessary in the first place because the element types in values_as_tensor is more than just int64, string, float.

Reply:
This has now been changed